### DISCUSSION TITLE
SQLAlchemy Async: How to synchronise the columns of a declarative model with its underlying table after metadata reflection?

### DISCUSSION THREAD

**Discussion by dcs3spp** [sqlalchemy/sqlalchemy#8981] [Usage Questions] [answered]
Hi,

I am using alembic to apply a migration for a postgresql view using an async engine. This is successfully applied to the database.

I have the following declarative mapped class to  the view, defined as:

``` python
class MailingListView(Base): 
    """View for mailing labels.
    After metata reflection from db -> model expecting columns to
    be available on this class. 
    """
    __tablename__ = "mailing_list_view"

    # Specify the column override from the underlying view that is the primary key
    id = Column(UUID(as_uuid=True), primary_key=True)
    
    # Expecting  these columns  below to be mapped in this class after
    # metadata reflection. Currently have to uncomment these
    # to manually synchronise with view!
    #
    # addressee = Column(String)
    # street = Column(String)
    # town = Column(String)
    # county = Column(String)
    # postcode = Column(String)
    # cursor = Column(String)
```

I am reflecting the views using the following:

```python
      def use_inspector(conn):
            inspector = inspect(conn)
            return inspector.get_view_names()

        views = await connection.run_sync(use_inspector)

        # I can see the table columns in __table__.c.keys()
        # after the reflection below has run
        await connection.run_sync(
            target_metadata.reflect,
            only=views,
            views=True,
            extend_existing=True,
        )
```

After applying migrations and performing the above reflection I can see that my mapped model has the underlying table columns updated with those defined in the underlying view.

```python
obj = MailingListView()
obj.__table__.c.keys()
```

However, the properties of my mapped class are not updated after reflection, raising an exception:

```
obj = MailingListView()
obj.town = "town" # this raises an exception with unavailable property
```

How is it possible for a postgresql db (asyncpg) + async sqlalchemy to:

*Synchronise the columns of a declarative model with its underlying table after metadata reflection?*

Currently, I have to manually specify the columns in the declarative model. Is there another way?

**Comment by zzzeek** [comment, marked as answer]
if you want mapped classes to include reflected table columns, you have to reflect those columns before the class is mapped.

If using inline declarative,  in order to achieve this you need to use [DeferredReflection](https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#using-deferredreflection).  overall the section at https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#mapping-declaratively-with-reflected-tables illustrates the various options for mixing reflection with mapping.

**Reply by dcs3spp** [reply]
Hi many thanks @zzzeek

Is ```DeferredReflection.prepare``` and ```autoload_with``` compatible with an Async engine?

**Reply by zzzeek** [reply]
sure, use [run_sync](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#session-run-sync) as always

**Reply by dcs3spp** [reply]
Thanks @zzzeek , ok I'll give the`DeferredReflection` method a try and will see what happens. Thanks again

**Reply by dcs3spp** [reply]
```python
class Reflected(DeferredReflection):
    """Mark declarative class to be reflected later

    The reflection is deferred until
    DeferredReflection.prepare() is called.
    See
    https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#using-deferredreflection
    for further details.
    """
    __abstract__ = True
```

```python
class MailingListView(Reflected, Base):
    __tablename__ = "mailing_list_view"
    
    # Specify the column from the underlying view that is the primary key
    id = Column(UUID(as_uuid=True), primary_key=True)
```

Now run the reflection and afterwards use `Reflected` prepare method

```
def use_inspector(conn):
            inspector = inspect(conn)
            return inspector.get_view_names()

views = await connection.run_sync(use_inspector)

# I can see the table columns in __table__.c.keys()
# after the reflection below has run
await connection.run_sync(
    target_metadata.reflect,
    only=views,
    views=True,
    extend_existing=True,
)

async_engine: AsyncEngine = connectable
await connection.run_sync(Reflected.prepare(async_engine))
```

This throws error:
```bash
sqlalchemy.exc.NoInspectionAvailable: Inspection on an AsyncEngine is currently not supported. Please obtain a connection then use ``conn.run_sync`` to pass a callable where it's possible to call ``inspect`` on the passed connection.
```

How to use `DeferredReflection prepare` with an `AsyncEngine` instance?

**Reply by zzzeek** [reply]
`await connection.run_sync(Reflected.prepare)` - the connection is passed as first argument

**Reply by dcs3spp** [reply]
Many thanks @zzzeek much appreciated!!!! That did the trick, it's working now!!!!
